crates/pipeline_manager/src/db/mod.rs

```diff
@@ -717,8 +989,154 @@ impl Storage for ProjectDB {
+        // Need to revise the API a bit to optionally pass in a transaction
+        // object for this to work seamlessly. Also probably the program row
+        // should be locked with FOR UPDATE while reading to prevent the
+        // compiler from resetting the status field.
+        let (_pipeline, program, _connectors) =
+            self.pipeline_is_committable(tenant_id, pipeline_id).await?;
+
+        // Find the revision number
+        let revision_data = txn
+            .query_opt(
+                "SELECT last_revision FROM pipeline WHERE id = $1 AND tenant_id = $2 FOR UPDATE",
+                &[&pipeline_id.0, &tenant_id.0],
+            )
+            .await?;
+        let (revision, prev_revision) = match revision_data {
+            Some(row) => {
+                let revision: i64 = row.get::<_, Option<i64>>(0).unwrap_or(0);
+                (revision + 1, revision)
+            }
+            None => return Err(DBError::UnknownPipeline(pipeline_id).into()),
+        };
+
+        // Check if we actually changed something before writing a new revision
+        //
+        // Note: What fields are checked is ultimately determined by whatever is
+        // used by the pipeline configuration, e.g., what the `start` function
+        // uses in `runner.rs` to write the config/metadata:
+        let change = txn.query_one(
+            "SELECT count(*) FROM
+            (
+                SELECT progh.code, ch.config, ach.name, ach.config, ach.is_input, ph.config
+                FROM pipeline_history ph, program_history progh, attached_connector_history ach, connector_history ch
+                WHERE ph.id = $1
+                    AND ph.program_id = progh.id
+                    AND ach.pipeline_id = ph.id
+                    AND ach.connector_id = ch.id
+                    AND ach.revision = $2
+                    AND progh.revision = $2
+                    AND ph.revision = $2
+                    AND ch.revision = $2
+                EXCEPT
+                SELECT prog.code, c.config, ac.name, ac.config, ac.is_input, p.config
+                FROM pipeline p, program prog, attached_connector ac, connector c
+                WHERE p.id = $1
+                    AND p.program_id = prog.id
+                    AND ac.pipeline_id = p.id
+                    AND ac.connector_id = c.id
+            ) as x",
+            &[&pipeline_id.0, &prev_revision],
+        )
+        .await?;
+        let nothing_changed = change.get::<_, i64>(0) == 0;
+        let committed_before = prev_revision > 0;
+        if committed_before && nothing_changed {
+            return Err(DBError::RevisionNotChanged.into());
+        }
+
+        // Copy all pipeline data to history tables
+        //
+        // TODO(performance): In theory the following inserts could all run in
+        // parallel with async but I couldn't figure out how to make it work
+        // with the args :/
+        txn.execute(
```

Comment:
If they can all be ran independently, maybe try future::try_join(f1, f2, f3...).await to execute them in parallel? Although I doubt they'll truly be executed in parallel given the underlying DB connection, but they might at least be pipelined.

Reply:
I tried that for a while but run into errors with arguments not living long enough -- which in turn I tried to solve by making them closure where it complained about different types... I can give it another shot

Reply:
I changed it to

```rust
        let version_program = async {
            txn.execute(
                "INSERT INTO program_history SELECT $1 as revision, * FROM program p WHERE id = $2",
                &[&revision, &program.program_id.0],
            )
            .await
        };
        let version_pipeline = async {
            txn.execute(
                "INSERT INTO pipeline_history SELECT $1 as revision, * FROM pipeline p WHERE id = $2",
                &[&revision, &pipeline_id.0],
            ).await
        };
        let version_connectors = async {
            txn.execute(
                "INSERT INTO connector_history SELECT $1 as revision, c.* FROM connector c, attached_connector ac WHERE ac.pipeline_id = $2 AND ac.connector_id = c.id",
                &[&revision, &pipeline_id.0],
            )
            .await
        };
        let version_acs = async {
            txn.execute(
                "INSERT INTO attached_connector_history SELECT $1 as revision, * FROM attached_connector ac WHERE pipeline_id = $2",
                &[&revision, &pipeline_id.0],
            ).await
        };
        let increment_revision = async {
            // Update the revision of the pipeline object
            txn.execute(
                "UPDATE pipeline SET last_revision = $1 WHERE id = $2 AND tenant_id = $3",
                &[&revision, &pipeline_id.0, &tenant_id.0],
            )
            .await
        };

        try_join!(
            version_program,
            version_pipeline,
            version_connectors,
            version_acs,
            q5
        )?;
```

which got the test to fail with

```
 value: db error: ERROR: current transaction is aborted, commands ignored until end of transaction block
```

so I'm giving up for now. I think the more pressing issue will be to convert this to prepared statements first anyways that will save one extra roundtrip every time.